### ISSUE TITLE
Proposal: intrinsic-size CSS property

### ISSUE THREAD

**Issue by chrishtr** [w3c/csswg-drafts#4229]
http://tabatkins.github.io/specs/css-content-size/

This property specifies the size to reserve for elements which already have `contain:size`. `contain:size` without `content-size` is like having `content-size: 0px 0px. It's useful for providing placeholder sizes of subtrees in cases when it is not desirable to render the subtrees.

See also https://github.com/WICG/display-locking/blob/master/explainer-new.md for examples and background that helps to motivate this property.

Feedback appreciated!

**Comment by dbaron** [comment]
We've talked in the past about a property that sets the intrinsic size(s) of an element.  Should this also be that property (or one of those properties) as well?

**Comment by chrishtr** [comment]
The standalone motivation for the property is to provide a placeholder sizing for subtree content which is either not yet available or not rendered.

**Comment by tabatkins** [comment]
The intrinsic size stuff's main motivation is getting a good size for a replaced element as soon it's in the document, rather than waiting for the resource to load. While using CSS for that has been discussed, it still puts the necessary information behind a resource load (the CSS stylesheet). So the current proposals for that have all stayed within HTML attributes so far, which seems reasonable.

So I lean towards not trying to merge these pieces of functionality, even tho I agree they're close in their natures.

**Comment by astearns** [comment]
I'm a little hesitant to add a separate property that only applies in particular `contain` scenarios. Could the size values be added as optional additional values to go along with the `size` value in `contain`? Something like `contain: size 200px 200px` or `contain: size(200px 200px)` might make the connection more clear.

**Comment by tabatkins** [comment]
We thought about that, but it ends up working awkwardly with our current async-render design.

Basically, once you tell a subtree to be async-rendered, it automatically applies several containments, in addition to whatever else you might be specifying in your own stylesheet. One of those is size containment, so if you needed to apply it yourself to get this functionality, that wouldn't work.

* Our earlier design instead required *you* to apply the necessary containments to get async-rendering to work at all, but that seemed awkward and bad. It puts a non-obvious hurdle in the way of authors getting this to work ("why isn't my element rendering async? I set the attribute correctly!") and it's future-hostile if we add more containment options in the future that we want async stuff to rely on (suddenly all the old pages using async will just become sync, as they're lacking necessary containment values).
* We assign the necessary containments "magically" in an additive way, rather than using a UA stylesheet that could be overridden, because of the lessons we learned with the `hidden` attribute: doing it in a normal UA rule becomes trivial to accidentally override when applying your own lesser containments, while doing it in a UA-important rule means you can't apply *extra* containments beyond what the feature requires, because the author rule will lose. If/when we get additive cascade features we can probably avoid the magic.
* We might be able to get around this by specifying that if both `size` containment and `size-with-child` containment are applied, the size-with-child wins out. Or that the magic doesn't attempt to add size containment if it detects size-with-child containment already in place.
* If the containment value definitely turns on size containment, then you need to make sure you only apply it when the element is invisible-subtree mode (`rendering="invisible"` or `rendering="invisible-activatable"`). That's not difficult (we purposely designed the feature to make the attribute a source of truth, so you can use it in selectors to style differently based on current state), but it is one more detail to think about, versus just blind applying a property that'll only have an effect when it matters. Alternately we could make it a containment value that just modifies size containment but has no effect on its own, which feels a *little* weird but I guess not terrible.

So, yeah, we've avoided it for now, but it's possible to do if the group decides a specialty property is overkill.

**Comment by vmpstr** [comment]
It would be nice for the spec to explicitly call out layout modes where the contain: size layout is not empty. The two examples I know of are:
- `<select multiple>`, that lays out as if it had a default number of default height items. My proposal is if content-size is not none, then it should lay out with all of its contents being sized to content-size (not just a single child, which would be the case with the current wording)
- grid layout, which may be sized to something based on various grid parameters. It's unclear whether it's a good idea to pretend we have a child, place it, and size based on that.

**Comment by chrishtr** [comment]
Here is a dedicated explainer for content-size, with extra examples:

https://github.com/WICG/display-locking/blob/master/explainer-content-size.md

**Comment by tabatkins** [comment]
Reopening for review - @dbaron, you had some significant thoughts about this property during TPAC. Mind giving it a look? <https://drafts.csswg.org/css-sizing-4/#intrinsic-contribution>

**Comment by tabatkins** [comment]
And agenda+ just to bring it up for review at the next call.

**Comment by Malvoz** [comment]
With the potential addition of [the `size` property](https://github.com/w3c/csswg-drafts/issues/820), I presume `proxy-size` should be a thing? It's probably too early to conclude since the `size` property hasn't been officially accepted, but I thought I'd mention it.

**Comment by tabatkins** [comment]
It's already in there (as `intrinsic-size`); you might have missed it because it's at the end of the section.

**Comment by heycam** [comment]
Is legacy really the right name for the initial value, i.e. is it always the objectively wrong thing to do for a scrollable element compared to auto, and it's just that we're unable to change the initial value due to compat?  Or could there be legitimate cases for using it?

I'd prefer if we had value names that indicated it is changing something about scrollable elements somehow, but I don't have any great suggestions.

**Comment by AmeliaBR** [comment]
> I'd prefer if we had value names that indicated it is changing something

What about `from-content` for `auto`, and then something like `from-content-including-overflow` for `legacy`? Make it clear what they mean, but also make it clear that the second one is an awkward thing that we don't usually recommend…

PS, Should we change the title of this issue to match the current name of the proposed property (`intrinsic-size`)?

**Comment by dbaron** [comment]
(I got here via w3ctag/design-reviews#437.)

I reviewed the [current definitions of the `intrinsic-*` properties](https://drafts.csswg.org/css-sizing-4/#intrinsic-size-override), and I'm somewhat concerned about the naming here given the way this interacts with `overflow`.

The current default behavior for the scrollable values of `overflow` (which include `hidden`) is that they lead to the min-content size being zeroed out (as far as the parent sees), but the max-content size propagating through the element to its parent.  In other words, this is a "middle" behavior in between propagating the intrinsic size through for both min-content and max-content, and zeroing out for both min-content and max-content.  (At least, I believe this is what Gecko implements, and I believe it had to implement it this way for compatibility.)

I recall seeing strong use cases for both behaviors for min-content size (zero out, or propagate through).  I'm less sure about for max-content size.  This makes me hesitant to have the default behavior called `legacy`; it is a behavior for which a good set of use cases exist.

I suspect (though I'm not sure) the non-default `auto` behavior is actually the less interesting deviation from the current default; I think there may be stronger use cases for "don't zero out min-content" than for "zero out max-content" (particularly with `overflow: hidden`).  At the very least, I'm pretty sure I recall seeing such use cases, but it's been a few years so I don't remember details.  I think we should offer this option.

Given that I think there are three reasonable options here for `overflow` interaction rather than two (and they should probably all be offered), I think `legacy` and `auto` are not good names.

(Also, if any of the three options *shouldn't* be offered because it doesn't have strong use cases, the one I would skip is the one currently defined as `auto`.  This is both because of the three I recall seeing the fewest use cases that needed it, and because it is largely achievable by specifying `0` (although only for certain `overflow` values).)

-----

I also think the definitions that refer to `overflow: visible` also should also include `overflow: clip`; I think the way to do that is to refer to whether the box establishes a [scroll container](https://drafts.csswg.org/css-overflow-3/#scroll-container).
